qcore/geo.py

```diff
@@ -995,16 +993,40 @@ def project_point_onto_plane(
     -------
     np.ndarray
         The projected points.
-
-    Examples
-    --------
-    FIXME: Add docs.
-
     """
-
-    return points - (
-        (points @ plane_dual_coordinates[:3]) - plane_dual_coordinates[3]
-    ).reshape((-1, 1)) @ plane_dual_coordinates[:3].reshape((1, -1))
+    # The point-normal description of a plane says a plane with normal vector
+    # n contains points p such that:
+    #
+    # n * p = 0.
+    #
+    # For affine planes (those translated away from the origin), we instead say
+    #
+    # n * p = d,
+    #
+    # where d is a translation distance from the origin. We can encapsulate
+    # this in the following diagram
+    #
+    #        ╱╲    normal
+    #       ╱  ╲ ╱
+    #      ╱    ╳
+    #     ╱    ╱ ╲
+    #    ╱    ╱   ╲
+    #    ╲   ╱    ╱
+    #     ╲ ╱    ╱
+    #    ^ ╲    ╱
+    #   d ╱ ╲  ╱
+    #  v ╱   ╲╱
+    #   ·
+    # origin
+    #
+    # So then n * p gives the distance in units of n from the origin. To
+    # project a point p onto a given plane with points n * r = d, we subtract
+    # (n * p) - d units of n from p, shifting it into the plane.
+    return points - np.outer(  # p -
+        np.dot(points, plane_dual_coordinates[:3])
+        - plane_dual_coordinates[3],  # ((n * p) - d) *
+        plane_dual_coordinates[:3],  # n
+    )
```

Comment:
I find this bit confusing. `p` is on a plane with normal vector that satisfies `n * p = 0`.
Then for origin, `n * p =d`. Is `p` still on the plane, or `p` is now origin?

Not sure if it is conventional to use * for both dot and outer operation. 
Where is `r` from in `a given plane with points n * r = d`. 

Also, `plane_dual_coordinates` are never mentioned in the comments, suddenly used in the code.

Reply:
Typically, I wouldn't use an operator for either of them. I would just have $\mathbf{u}^T \mathbf{v}$ and $\mathbf{u} \mathbf{v}^T$ as the underlying operation is always matrix multiplication. Wikipedia uses $\otimes$ so maybe I could use that?

Reply:
I'll think of a good way to clarify how plane dual coordinates are used in the code.

Reply:
but broadly, the dual coordinates can be thought of in this case as the normal to the plane plus this "d" term. So `n = plane_dual_coordinates[:3]` and `d = plane_dual_coordinates[3]`

Reply:
The use of the outer product, in case this isn't clear, is to allow projection of multiple points onto the plane without iterating over them in a for loop.